docs/docs/tutorial/chapter3/saving-data.md

```diff
@@ -625,26 +702,516 @@ const ContactPage = () => {
 export default ContactPage
 ```
 
-![Toast notification on successful submission](https://user-images.githubusercontent.com/300/146271487-f6b77e76-99c1-43e8-bcda-5ba3c9b03137.png)
+</TabItem>
+</Tabs>
 
-You can read the full documentation for Toast [here](../../toast-notifications.md).
+`create` is a function that invokes the mutation and takes an object with a `variables` key, containing another object with an `input` key. As an example, we could call it like:
 
-### Displaying Server Errors
+```js
+create({
+  variables: {
+    input: {
+      name: 'Rob',
+      email: 'rob@redwoodjs.com',
+      message: 'I love Redwood!',
+    },
+  },
+})
+```
 
-Next we'll inform the user of any server errors. So far we've only notified the user of _client_ errors: a field was missing or formatted incorrectly. But if we have server-side constraints in place `<Form>` can't know about those, but we still need to let the user know something went wrong.
+If you'll recall `<Form>` gives us all of the fields in a nice object where the key is the name of the field, which means the `data` object we're receiving in `onSubmit` is already in the proper format that we need for the `input`!
 
-We have email validation on the client, but any developer worth their silicon knows [never trust the client](https://www.codebyamir.com/blog/never-trust-data-from-the-browser). Let's add the email validation into the api side as well to be sure no bad data gets into our database, even if someone somehow bypassed our client-side validation (l33t hackers do this all the time).
+That means we can update the `onSubmit` function to invoke the mutation with the data it receives:
 
-> **No server-side validation for some fields?**
->
-> Why don't we need server-side validation for the existence of name, email and message? Because GraphQL is already doing that for us! You may remember the `String!` declaration in our SDL file for the `Contact` type: that adds a constraint that those fields cannot be `null` as soon as it arrives on the api side. If it is, GraphQL would reject the request and throw an error back to us on the client.
+<Tabs groupId="js-ts">
+<TabItem value="js" label="JavaScript">
+
+```jsx title="web/src/pages/ContactPage/ContactPage.js"
+import { MetaTags, useMutation } from '@redwoodjs/web'
+import {
+  FieldError,
+  Form,
+  Label,
+  TextField,
+  TextAreaField,
+  Submit,
+  SubmitHandler,
+} from '@redwoodjs/forms'
+
+const CREATE_CONTACT = gql`
+  mutation CreateContactMutation($input: CreateContactInput!) {
+    createContact(input: $input) {
+      id
+    }
+  }
+`
+
+interface FormValues {
+  name: string
+  email: string
+  message: string
+}
+
+const ContactPage = () => {
+  const [create] = useMutation(CREATE_CONTACT)
```

Comment:
```suggestion
  const [create] = useMutation<CreateContactMutation>(CREATE_CONTACT)
```

Reply:
I added `CreateContactMutationVariables` here as well

Reply:
Think maybe you've forgotten to push!

Reply:
I did push, but the changes didn't go in here, but rather in the TS code block right below.